systems/plants/@Manipulator/Manipulator.m

```diff
@@ -36,6 +36,73 @@
       C_times_v = C-G;
     end
     
+    function [xdot,dxdot] = Hdynamics(obj,t,x,u)
+    % Provides the DrakeSystem interface to the manipulatorDynamics.
+    %the dynamics left multiplied by H
+
+      q = x(1:obj.num_positions);
+      v = x(obj.num_positions+1:end);
+
+      if (nargout>1)
+        if ~isempty(obj.position_constraints) || ~isempty(obj.velocity_constraints)
+          % by naming this 'MATLAB:TooManyOutputs', geval will catch the
+          % error and use TaylorVarInstead
+          error('MATLAB:TooManyOutputs','User gradients for constrained dynamics not implemented yet.');
+        end
+
+        % Note: the next line assumes that user gradients are implemented.
+        % If it fails, then it will raise the same exception that I would
+        % want to raise for this method, stating that not all outputs were
+        % assigned.  (since I can't write dxdot anymore)
+        [H,C,B,dH,dC,dB] = obj.manipulatorDynamics(q,v);
+        
+        %ANDY CHANGE
+        %Hinv = inv(H);
+        %Hinv = invmsspoly(H);
+
+        if (obj.num_u>0)
+          vdot = (B*u-C);
+          dtau = matGradMult(dB,u) - dC;
+          dvdot = [zeros(obj.num_positions,1),...
+            -matGradMult(dH(:,1:obj.num_positions),vdot) + dtau(:,1:obj.num_positions),...
+            +dtau(:,1+obj.num_positions:end), B];
+        else
+          vdot = -C;
+          dvdot = [zeros(obj.num_velocities,1),...
+            (-matGradMult(dH(:,1:obj.num_positions),vdot) - dC(:,1:obj.num_positions)),...
+            (-dC(:,obj.num_positions+1:end))];
+        end
+
+        [VqInv,dVqInv] = vToqdot(obj,q);
+        xdot = [VqInv*v;vdot];
+        dxdot = [...
+          zeros(obj.num_positions,1), matGradMult(dVqInv, v), VqInv, zeros(obj.num_positions,obj.num_u);
+          dvdot];
+      else
+        [H,C,B] = manipulatorDynamics(obj,q,v);
+        if (obj.num_u>0) tau=B*u - C; else tau=-C; end
+        tau = tau + computeConstraintForce(obj,q,v,H,tau,eye(size(H, 1)));
+
+        vdot = tau;
```

Comment:
It is possible that there is an error in it.  I wound up changing courses
and going the dynamics, symbolicized route.  If you would prefer, I can
remove the Hdynamics function.


On Thu, Jul 30, 2015 at 10:57 PM, Twan Koolen <notifications@github.com>
wrote:

> In systems/plants/@Manipulator/Manipulator.m
> <https://github.com/RobotLocomotion/drake/pull/1212#discussion_r35943267>:
>
> > +          dvdot = [zeros(obj.num_velocities,1),...
> > +            (-matGradMult(dH(:,1:obj.num_positions),vdot) - dC(:,1:obj.num_positions)),...
> > +            (-dC(:,obj.num_positions+1:end))];
> > +        end
> > +
> > +        [VqInv,dVqInv] = vToqdot(obj,q);
> > +        xdot = [VqInv*v;vdot];
> > +        dxdot = [...
> > +          zeros(obj.num_positions,1), matGradMult(dVqInv, v), VqInv, zeros(obj.num_positions,obj.num_u);
> > +          dvdot];
> > +      else
> > +        [H,C,B] = manipulatorDynamics(obj,q,v);
> > +        if (obj.num_u>0) tau=B*u - C; else tau=-C; end
> > +        tau = tau + computeConstraintForce(obj,q,v,H,tau,eye(size(H, 1)));
> > +
> > +        vdot = tau;
>
> This makes no sense.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/RobotLocomotion/drake/pull/1212/files#r35943267>.
>